### ISSUE TITLE
aws_codebuild_project: missing proper way to conditionally set vpc_config

### ISSUE THREAD

**Issue by ebarault** [terraform-providers/terraform-provider-aws#4115]
I'm building a `aws_codebuild_project` module where vpc integration is enabled/disabled conditionally.
Ideally, setting all the params in the `vpc_config` block to `""` or empty list would have worked to pass a deactivated vpc config, as in: 
```hcl
  vpc_config {
    vpc_id = ""
    subnets = []
    security_group_ids = []
  }
```
but it does not. Worst: applying such a config throws an error and breaks the remote state file and terraform further on proposes to create entirely the existing resource when reapplying after the resulting error.

I found the hacky `vpc_config = []` option to work, but I know it's not officially supported, and hence is fragile.

### Expected Behavior
A robust way should be provided to configure to deactivate codebuild project's vpc config while setting the `vpc_config` block inside the module code

### Terraform Version
0.11.6

### AWS provider version
1.14.0

### Affected Resource(s)
- aws_codebuild_project

### Terraform Configuration Files
```hcl
resource "aws_codebuild_project" "codebuild_project" {
  # ...

  # working config
  vpc_config = []

  # non working config
  vpc_config {
    vpc_id = ""
    subnets = []
    security_group_ids = []
  }

}
```

cc: @drsolaris, @bflad

**Comment by rrati** [comment]
Could you do this with count?  Set count to 0 if your enable condition is false

**Comment by ebarault** [comment]
@rrati : i'm not intending to conditionally enable/disable all the codebuild project module, just the vpc configuration part. Hence using count would not work here

**Comment by catsby** [comment]
>I found the hacky vpc_config = [] option to work, but I know it's not officially supported, and hence is fragile

Forgive my ignorance, where is it said this is not supported? I agree it's a bit hacky ;)

**Comment by ebarault** [comment]
well, 

**1.**: assigning
```
some_block {
}
```
using
```
some__block = [{
}]
```
is documented nowhere although it is used a lot when building advanced modules

**2.** its support varies from module to module, sometimes half working, sometimes throwing exceptions. Take for example the `cache` block in this very same module: assigning 
```
cache = []
```
does work in the sense that it ends up not enabling cache, but it only half works as it yet does trigger some changes when applying the module

`3.` several similar comments from @apparentlymart, including this [one](https://github.com/hashicorp/terraform/issues/16959#issuecomment-355767964)

> due to some coincidences in Terraform's implementation today you can sometimes trick Terraform into treating a nested block as an attribute by assigning a list of maps that happens to have the same internal representation as is currently used for nested blocks (a Go map[string]interface{}), but this is not an officially-supported feature and, as we see here, does not work robustly.

**Comment by catsby** [comment]
Thanks! I'll defer to @apparentlymart for more comment here(?). I know we have some upcoming enhancements to HCL that may help here, but I can't promise. Sorry I don't have a better answer for you 😞

**Comment by ebarault** [comment]
@catsby, @apparentlymart : any chance a fix could be worked out independently of the new forthcoming HCL enhancements?

Shouldn't `""` values be interpreted by terraform as `disabled` by default, as in most modules?

**Comment by araddas** [comment]
I am getting around this by creating a map similar to this, then pass a boolean  to var.function_vpc_enable to access the map key. It seems to work as it should, however I just ran into an issue with the  aws_subnet_ids data resource, if I pass the list from that resource it fails, but a hardcoded list works as it should.
```
vpcconfig  = {
    "1" = [{
        subnet_ids         = "${var.function_vpc_subnet_ids}",
        security_group_ids = "${var.function_vpc_sg_id}"
     }]
    "0" = []
  }
  
resource "aws_lambda_function" "lambda-function" {
  ...
  ...
  vpc_config  = "${local.vpcconfig[var.function_vpc_enable]}"
}
```

**Comment by bflad** [comment]
Hi folks 👋 This issue is resolved in Terraform 0.12, which supports new functionality in the configuration language aimed at solving problems like these. The new [`dynamic` block syntax](https://www.terraform.io/docs/configuration/expressions.html#dynamic-blocks) can be used to dynamically generate configuration blocks and their arguments. These can be combined with the new [`null` value](https://www.terraform.io/docs/configuration/expressions.html#null), which can be used to omit arguments as if they were not defined in the configuration at all.

I was able to implement a conditional `vpc_config` configuration block based on a variable with the below configuration:

```hcl
terraform {
  required_providers {
    aws = "2.21.1"
  }
  required_version = "0.12.5"
}

provider "aws" {
  region = "us-east-2"
}

variable "vpc_enabled" {
  default = true
  type    = bool
}

data "aws_availability_zones" "available" {
  # InvalidInputException: CodeBuild currently doesn't support VPC in us-west-2d, please select subnets in other availability zones.
  blacklisted_zone_ids = ["usw2-az4"]
  state                = "available"
}

resource "aws_vpc" "test" {
  count = var.vpc_enabled ? 1 : 0

  cidr_block = "10.0.0.0/16"
}

resource "aws_security_group" "test" {
  count = var.vpc_enabled ? 1 : 0

  vpc_id = aws_vpc.test[0].id
}

resource "aws_subnet" "test" {
  count = var.vpc_enabled ? 1 : 0

  availability_zone = data.aws_availability_zones.available.names[count.index]
  cidr_block        = "10.0.${count.index}.0/24"
  vpc_id            = aws_vpc.test[0].id
}

locals {
  vpc_config = var.vpc_enabled ? [{
    security_group_ids = aws_security_group.test[*].id
    subnets            = aws_subnet.test[*].id
    vpc_id             = aws_vpc.test[0].id
  }] : []
}

resource "aws_iam_role" "test" {
  assume_role_policy = jsonencode({
    Version = "2012-10-17"
    Statement = [{
      Action = "sts:AssumeRole"
      Effect = "Allow",
      Principal = {
        Service = "codebuild.amazonaws.com"
      }
    }]
  })
}

resource "aws_iam_role_policy" "test" {
  role = aws_iam_role.test.name

  policy = jsonencode({
    Version = "2012-10-17",
    Statement = [
      {
        Action = [
          "logs:CreateLogGroup",
          "logs:CreateLogStream",
          "logs:PutLogEvents",
        ]
        Effect   = "Allow",
        Resource = ["*"]
      },
      {
        Action = [
          "ec2:CreateNetworkInterface",
          "ec2:CreateNetworkInterfacePermission",
          "ec2:DescribeDhcpOptions",
          "ec2:DescribeNetworkInterfaces",
          "ec2:DeleteNetworkInterface",
          "ec2:DescribeSubnets",
          "ec2:DescribeSecurityGroups",
          "ec2:DescribeVpcs",
        ]
        Effect   = "Allow"
        Resource = "*"
      },
      {
        Action   = ["s3:*"],
        Effect   = "Allow",
        Resource = "*"
      },
    ]
  })
}

resource "aws_codebuild_project" "test" {
  name         = "test"
  service_role = "${aws_iam_role.test.arn}"

  artifacts {
    type = "NO_ARTIFACTS"
  }

  environment {
    compute_type = "BUILD_GENERAL1_SMALL"
    image        = "2"
    type         = "LINUX_CONTAINER"
  }

  source {
    location = "https://github.com/hashicorp/packer.git"
    type     = "GITHUB"
  }

  dynamic "vpc_config" {
    for_each = local.vpc_config

    content {
      security_group_ids = vpc_config.value.security_group_ids
      subnets            = vpc_config.value.subnets
      vpc_id             = vpc_config.value.vpc_id
    }
  }
}
```

When applying this:

```console
$ terraform apply
data.aws_availability_zones.available: Refreshing state...

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_codebuild_project.test will be created
  + resource "aws_codebuild_project" "test" {
      + arn            = (known after apply)
      + badge_enabled  = false
      + badge_url      = (known after apply)
      + build_timeout  = 60
      + description    = (known after apply)
      + encryption_key = (known after apply)
      + id             = (known after apply)
      + name           = "test"
      + service_role   = (known after apply)

      + artifacts {
          + encryption_disabled = false
          + type                = "NO_ARTIFACTS"
        }

      + environment {
          + compute_type                = "BUILD_GENERAL1_SMALL"
          + image                       = "2"
          + image_pull_credentials_type = "CODEBUILD"
          + privileged_mode             = false
          + type                        = "LINUX_CONTAINER"

          + environment_variable {
              + name  = (known after apply)
              + type  = (known after apply)
              + value = (known after apply)
            }
        }

      + source {
          + location = "https://github.com/hashicorp/packer.git"
          + type     = "GITHUB"
        }

      + vpc_config {
          + security_group_ids = (known after apply)
          + subnets            = (known after apply)
          + vpc_id             = (known after apply)
        }
    }

  # aws_iam_role.test will be created
  + resource "aws_iam_role" "test" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + Service = "codebuild.amazonaws.com"
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + max_session_duration  = 3600
      + name                  = (known after apply)
      + path                  = "/"
      + unique_id             = (known after apply)
    }

  # aws_iam_role_policy.test will be created
  + resource "aws_iam_role_policy" "test" {
      + id     = (known after apply)
      + name   = (known after apply)
      + policy = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = [
                          + "logs:CreateLogGroup",
                          + "logs:CreateLogStream",
                          + "logs:PutLogEvents",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "*",
                        ]
                    },
                  + {
                      + Action   = [
                          + "ec2:CreateNetworkInterface",
                          + "ec2:CreateNetworkInterfacePermission",
                          + "ec2:DescribeDhcpOptions",
                          + "ec2:DescribeNetworkInterfaces",
                          + "ec2:DeleteNetworkInterface",
                          + "ec2:DescribeSubnets",
                          + "ec2:DescribeSecurityGroups",
                          + "ec2:DescribeVpcs",
                        ]
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                  + {
                      + Action   = [
                          + "s3:*",
                        ]
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + role   = (known after apply)
    }

  # aws_security_group.test[0] will be created
  + resource "aws_security_group" "test" {
      + arn                    = (known after apply)
      + description            = "Managed by Terraform"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + vpc_id                 = (known after apply)
    }

  # aws_subnet.test[0] will be created
  + resource "aws_subnet" "test" {
      + arn                             = (known after apply)
      + assign_ipv6_address_on_creation = false
      + availability_zone               = "us-east-2a"
      + availability_zone_id            = (known after apply)
      + cidr_block                      = "10.0.0.0/24"
      + id                              = (known after apply)
      + ipv6_cidr_block                 = (known after apply)
      + ipv6_cidr_block_association_id  = (known after apply)
      + map_public_ip_on_launch         = false
      + owner_id                        = (known after apply)
      + vpc_id                          = (known after apply)
    }

  # aws_vpc.test[0] will be created
  + resource "aws_vpc" "test" {
      + arn                              = (known after apply)
      + assign_generated_ipv6_cidr_block = false
      + cidr_block                       = "10.0.0.0/16"
      + default_network_acl_id           = (known after apply)
      + default_route_table_id           = (known after apply)
      + default_security_group_id        = (known after apply)
      + dhcp_options_id                  = (known after apply)
      + enable_classiclink               = (known after apply)
      + enable_classiclink_dns_support   = (known after apply)
      + enable_dns_hostnames             = (known after apply)
      + enable_dns_support               = true
      + id                               = (known after apply)
      + instance_tenancy                 = "default"
      + ipv6_association_id              = (known after apply)
      + ipv6_cidr_block                  = (known after apply)
      + main_route_table_id              = (known after apply)
      + owner_id                         = (known after apply)
    }

Plan: 6 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_vpc.test[0]: Creating...
aws_iam_role.test: Creating...
aws_iam_role.test: Creation complete after 1s [id=terraform-20190731022106333500000001]
aws_iam_role_policy.test: Creating...
aws_iam_role_policy.test: Creation complete after 0s [id=terraform-20190731022106333500000001:terraform-20190731022106659200000002]
aws_vpc.test[0]: Creation complete after 2s [id=vpc-04643a2db846dba99]
aws_subnet.test[0]: Creating...
aws_security_group.test[0]: Creating...
aws_subnet.test[0]: Creation complete after 1s [id=subnet-0abbabdb080fea021]
aws_security_group.test[0]: Creation complete after 1s [id=sg-0b1da906a92be6c97]
aws_codebuild_project.test: Creating...
aws_codebuild_project.test: Creation complete after 9s [id=arn:aws:codebuild:us-east-2:--OMITTED--:project/test]

Apply complete! Resources: 6 added, 0 changed, 0 destroyed.

$ terraform state show aws_codebuild_project.test
# aws_codebuild_project.test:
resource "aws_codebuild_project" "test" {
    arn            = "arn:aws:codebuild:us-east-2:--OMITTED--:project/test"
    badge_enabled  = false
    build_timeout  = 60
    encryption_key = "arn:aws:kms:us-east-2:--OMITTED--:alias/aws/s3"
    id             = "arn:aws:codebuild:us-east-2:--OMITTED--:project/test"
    name           = "test"
    service_role   = "arn:aws:iam::--OMITTED--:role/terraform-20190731022106333500000001"

    artifacts {
        encryption_disabled = false
        type                = "NO_ARTIFACTS"
    }

    cache {
        modes = []
        type  = "NO_CACHE"
    }

    environment {
        compute_type                = "BUILD_GENERAL1_SMALL"
        image                       = "2"
        image_pull_credentials_type = "CODEBUILD"
        privileged_mode             = false
        type                        = "LINUX_CONTAINER"
    }

    logs_config {
        cloudwatch_logs {
            status = "ENABLED"
        }

        s3_logs {
            encryption_disabled = false
            status              = "DISABLED"
        }
    }

    source {
        location = "https://github.com/hashicorp/packer.git"
        type     = "GITHUB"
    }

    vpc_config {
        security_group_ids = [
            "sg-0b1da906a92be6c97",
        ]
        subnets            = [
            "subnet-0abbabdb080fea021",
        ]
        vpc_id             = "vpc-04643a2db846dba99"
    }
}
```

Setting the variable `vpc_enabled` to `false` yields a completely missing VPC infrastructure and config within the CodeBuild Project:

```console
$ terraform apply
data.aws_availability_zones.available: Refreshing state...

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_codebuild_project.test will be created
  + resource "aws_codebuild_project" "test" {
      + arn            = (known after apply)
      + badge_enabled  = false
      + badge_url      = (known after apply)
      + build_timeout  = 60
      + description    = (known after apply)
      + encryption_key = (known after apply)
      + id             = (known after apply)
      + name           = "test"
      + service_role   = (known after apply)

      + artifacts {
          + encryption_disabled = false
          + type                = "NO_ARTIFACTS"
        }

      + environment {
          + compute_type                = "BUILD_GENERAL1_SMALL"
          + image                       = "2"
          + image_pull_credentials_type = "CODEBUILD"
          + privileged_mode             = false
          + type                        = "LINUX_CONTAINER"

          + environment_variable {
              + name  = (known after apply)
              + type  = (known after apply)
              + value = (known after apply)
            }
        }

      + source {
          + location = "https://github.com/hashicorp/packer.git"
          + type     = "GITHUB"
        }
    }

  # aws_iam_role.test will be created
  + resource "aws_iam_role" "test" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + Service = "codebuild.amazonaws.com"
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + max_session_duration  = 3600
      + name                  = (known after apply)
      + path                  = "/"
      + unique_id             = (known after apply)
    }

  # aws_iam_role_policy.test will be created
  + resource "aws_iam_role_policy" "test" {
      + id     = (known after apply)
      + name   = (known after apply)
      + policy = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = [
                          + "logs:CreateLogGroup",
                          + "logs:CreateLogStream",
                          + "logs:PutLogEvents",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "*",
                        ]
                    },
                  + {
                      + Action   = [
                          + "ec2:CreateNetworkInterface",
                          + "ec2:CreateNetworkInterfacePermission",
                          + "ec2:DescribeDhcpOptions",
                          + "ec2:DescribeNetworkInterfaces",
                          + "ec2:DeleteNetworkInterface",
                          + "ec2:DescribeSubnets",
                          + "ec2:DescribeSecurityGroups",
                          + "ec2:DescribeVpcs",
                        ]
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                  + {
                      + Action   = [
                          + "s3:*",
                        ]
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + role   = (known after apply)
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_iam_role.test: Creating...
aws_iam_role.test: Creation complete after 1s [id=terraform-20190731022213964000000001]
aws_iam_role_policy.test: Creating...
aws_codebuild_project.test: Creating...
aws_iam_role_policy.test: Creation complete after 0s [id=terraform-20190731022213964000000001:terraform-20190731022214662800000002]
aws_codebuild_project.test: Creation complete after 9s [id=arn:aws:codebuild:us-east-2:--OMITTED--:project/test]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
$ terraform state show aws_codebuild_project.test
# aws_codebuild_project.test:
resource "aws_codebuild_project" "test" {
    arn            = "arn:aws:codebuild:us-east-2:--OMITTED--:project/test"
    badge_enabled  = false
    build_timeout  = 60
    encryption_key = "arn:aws:kms:us-east-2:--OMITTED--:alias/aws/s3"
    id             = "arn:aws:codebuild:us-east-2:--OMITTED--:project/test"
    name           = "test"
    service_role   = "arn:aws:iam::--OMITTED--:role/terraform-20190731022213964000000001"

    artifacts {
        encryption_disabled = false
        type                = "NO_ARTIFACTS"
    }

    cache {
        modes = []
        type  = "NO_CACHE"
    }

    environment {
        compute_type                = "BUILD_GENERAL1_SMALL"
        image                       = "2"
        image_pull_credentials_type = "CODEBUILD"
        privileged_mode             = false
        type                        = "LINUX_CONTAINER"
    }

    logs_config {
        cloudwatch_logs {
            status = "ENABLED"
        }

        s3_logs {
            encryption_disabled = false
            status              = "DISABLED"
        }
    }

    source {
        location = "https://github.com/hashicorp/packer.git"
        type     = "GITHUB"
    }
}
```

Enjoy! 🚀
